An export goal for exporting data to be read by IDEs.

benjyw opened the pull request:
This change provides:
    
1) A generic extensible mechanism for backends to add exportable data.

2) An implementation in the python backend that exports a venv with 
all the target set's 3rdparty deps.

Follow-up changes can do things like export codegenned sources, or export
source roots into formats that PyCharm and VSCode can read.

[ci skip-rust]
    
[ci skip-build-wheels]

jsirois commented:
> One side-effect of this current dir structure is the virtual environment thinks it's name is (ddab8011daaee380698ac2fb9701af18c90c03f6.19d82559650744319baba2979601e291)

Hrm. This is because Benjy is re-using auto-created venvs never intended to be directly used. You can also created a venv for direct use (sourcing activate, mutating with pip, etc) via `pex-tools <the pex> venv <venv dir to create>`. If that were used, then the basename of `<venv dir to create>` would be your `(...)` prompt contents. That, though, is slow - up to O(seconds) if the venv is very large. IOW, Benjy uis using the right approach here I think. It looks like Pex could pass `--prompt` to both virtualenv (covering Python 2.7) and  to `python -mvenv` (covering Python 3.6+). The only hole is Python 3.5 which has a venv module that does not support this customization. I could either add support for and then expose that option in Pex or else just have Pex create auto-venvs with the prompt named after the base name of the PEX file by default.

benjyw commented:
Yeah, I'm using those venvs for performance reasons. 

Having Pex always set the prompt to something readable would be great. And if a user happens to be using Python 3.5, then apologies...

jsirois commented:
> Having Pex always set the prompt to something readable would be great. And if a user happens to be using Python 3.5, then apologies...

https://github.com/pantsbuild/pex/pull/1499

benjyw commented:
This is ready for initial review, thanks.

jsirois commented on src/python/pants/backend/python/goals/export.py:
```diff
@@ -0,0 +1,82 @@
+        raise ExportError(
+            "There is no single Python interpreter compatible with all the "
+            "targets for which export was requested. Please restrict the target set "
+            "to one that shares a compatible interpreter."
```
It would probably be helpful to include the value of `interpreter_constraints` in the message.

benjyw replied:
Done.

jsirois commented on src/python/pants/backend/python/goals/export.py:
```diff
@@ -0,0 +1,82 @@
+    complete_pex_env = pex_env.in_workspace()
+    venv_abspath = os.path.join(complete_pex_env.pex_root, venv_pex.venv_rel_dir)
+
+    # Run the venv_pex to ensure that the underlying venv is created if necessary.
```
IIUC that is guaranteed to always happen above (it's an important part of perf and why `--seed` was introduced and used); so maybe drop this verbage and focus on the `-c` related verbage which you know is true locally.

benjyw replied:
Done.

jsirois commented on src/python/pants/backend/python/goals/export.py:
```diff
@@ -0,0 +1,82 @@
+            description="Create virtualenv",
+            argv=["-c", "import sys; print('.'.join(str(x) for x in sys.version_info[0:3]))"],
+            input_digest=venv_pex.digest,
+            # TODO: Is there always a python_configured?
```
You can kill this setting of `extra_env` completely - all env vars are sealed into the venv script when it's created: https://github.com/pantsbuild/pants/blob/5fcdd8ae6c0e2731767862561d6d885fb849fc3d/src/python/pants/backend/python/util_rules/pex.py#L708-L769

benjyw replied:
Done.

jsirois reviewed (approved):
Tests would be great, but this looks right to me.

kaos reviewed (approved):
Works for me. :)

Possible future feature that I would like is to also be able to get any 1st party distribution in "editable" mode in the venv..

jsirois commented:
> Possible future feature that I would like is to also be able to get any 1st party distribution in "editable" mode in the venv..

IIUC editable mode installs are just symlinks to live sources. Here we use copies and - per PR discussion - `./pants --loop ...` for liveness refreshing.

kaos commented:
> Here we use copies

For 3rd party code only. There's no 1st party code in this export (from what I could tell in my case, at least)

joshua-cannon-techlabs commented:
Re: First-party. That's something I didn't think of (but not required for _my_ use-case). If you're using source roots (E.g. `src/python/...`) your editor isn't likely to pick up on this without config. So either: A) You teach your editor about the source root, or B) export also symlinks the source roots in the shim "venv"

benjyw commented:
I was thinking of exporting IDE config that points to source roots, since the IDE (at least pycharm) will treat sources in the venv as third-party.

jsirois commented:
> I was thinking of exporting IDE config that points to source roots, since the IDE (at least pycharm) will treat sources in the venv as third-party.

That will work for simple cases with consistent global resolves which you already ~limit this too anyhow. For other cases we'll need link farms just containing the sources compatible with the given resolve.

stuhood commented on src/python/pants/backend/python/goals/export.py:
```diff
@@ -0,0 +1,82 @@
+    pass
+
+
+@rule
+async def export_venv(
+    request: ExportedVenvRequest, python_setup: PythonSetup, pex_env: PexEnvironment
+) -> ExportableData:
```
Rather than creating a virtualenv in the dest, this is symlinking from PEX's private cache of venvs into a destination... that seems maybe a bit dangerous, as someone who activated this venv and then fiddled with it would cause reproducibility issues by mutating PEX's cache.

Should this actually create a brand new venv instead, using `pex-tool`'s facilities for that?

jsirois replied:
This is a valid concern. We're walking a perf line here,

Note that the fiddling would have to be editing by hand since these venvs don't get Pip installed into them (unlike `PEX_TOOLS=1 ./my.pex venv here` venvs, which do by default.).

stuhood replied:
Mm... that does alleviate the concern quite a bit. Not a blocker then probably.

benjyw commented:
Addressed the comments and added tests. PTAL. 

In a follow up I can optionally change symlinking to copying, to give the user a tradeoff between safety and performance (with the default on the side of safety).

benjyw commented:
Please note the further changes to interpreter_constraints.py - the test actually exposed a problem there: if there were no targets with an interpreter_constraints field, we'd return an empty constraint, instead of the global constraints. Now we return the global constraints.

Eric-Arellano commented on src/python/pants/core/goals/export.py:
```diff
@@ -0,0 +1,124 @@
+
+
+class ExportSubsystem(GoalSubsystem):
+    name = "export"
```
What's the story with `export-codegen`? Should that still exist? I wonder if this should be something like `export-ide`?

benjyw replied:
We can probably get rid of `export-codegen` once `export` exports it. I don't want to call it `export-ide` because it may be useful for other purposes. And that is a clunky name. This can be an "export various things" goal, with IDEs being a good use-case, but perhaps not the only one.

joshua-cannon-techlabs replied:
I'm interested to see the evolution of `export`. Seems like it will have one-or-more subcommands for various (sub-?)subsystems.... sub-goals?. 

I'm interested in possibly adding functionality to the `export` goal for exporting a best-guess `module_mapping`.

benjyw replied:
How would this be consumed? That functionality would be awesome, but `export` may not be the best place for it.

joshua-cannon-techlabs replied:
By humans I suppose. You're probably right, let me switch hats: ... and I've forgotten my other use-case. Ignore me for now :smile:

thejcannon replied:
Documentation! `sphinx`-generated documentation. Does `pants` support this? I don't think I saw it in the docs.

benjyw replied:
It doesn't, although someone else asked about this recently: https://pantsbuild.slack.com/archives/C046T6T9U/p1635972032450400

benjyw replied:
A "docgen" plugin could invoke Sphinx for Python, javadoc for Java, scaladoc for Scala and so on. But I expect it would be its own goal, not part of `export`.

thejcannon replied:
Is there a good reason why it shouldn't be a subset of `export`? If `codegen` is a subset, why shouldn't `docgen` be as well?

Somewhere else someone (was it you?) brought up trying to keep the list of goals light to not have too much in `help`, and I kinda agree with that.

kaos replied:
I tend to agree, that `export` could be a generic goal for getting anything out of the pants world to disk.

benjyw replied:
Well, export is more about exporting internal information, for consumption by other tools, such as IDEs. We'd export codegen for that reason only - in normal use codegen is a build byproduct that is not visible to the end user. But since the user may want to manually inspect it, and may want to view it in their IDE, we'd export it. 

Documentation, though, is a build product, like a PEX file or a distribution or a docker image or whatever. It is an actual formal output of the system that you can request. So I don't think `export` is the right place for it, just as we don't use `export` to get compiled code, or any other build product.

stuhood replied:
Agree with @benjyw on this one... I expect that docs meet the bar for a dedicated goal. Likewise, I think that `export` being "for anything you want to get out of pants" is probably too broad: focusing on the IDE usecase is valid (i.e., making this page as short as possible: https://www.pantsbuild.org/docs/setting-up-an-ide).

thejcannon replied:
Yeah OK I agree. But now you have me wondering if we should `package` documentation :joy:

stuhood replied:
Good question!
